Bug 1253112 - Add support for atomic uploads for StorageClient

sleroux opened the pull request:
This PR adds iOS support for the new batch upload API for [this](https://bugzilla.mozilla.org/show_bug.cgi?id=1250189) bug. WIP documentation for the API can be found [here](https://github.com/mozilla-services/docs/pull/60/files).

sleroux commented:
@rnewman Figured I would open up a PR for consideration early to get your feedback on the code. The code tries to accomplish:

* A functional wrapping around the stateful nature of the new batch API. Consumers of a Sync15CollectionClient don't need to worry about keeping track of batch state as it is handled internally by the new `batchPost` API.

* Replacement for the existing 'chunk' uploading logic in `TimestampedSingleCollectionSynchronizer` with the new batch API call.

Things missing:

* Adding additional `X-Weave-*` data to the header to aid the server in failing fast. Not sure if we need this on iOS.

* The current implementation fails when no batch token is present from the batch start call. Not sure of the cleanest way of handling this. Probably fall back to the original implementation of chunk uploading?

* Tests...

rnewman commented:
My initial reaction to this is that it's inside out. (It also requires that the caller know the whole set of records up-front, which we don't always, so it doesn't quite work!)

For example: `LoginsSynchronizer` calls `getDeletedLoginsToUpload` and `getModifiedLoginsToUpload`, chaining each into the same storage client. With this change those two will be separate batches, which is inconsistent. History does the same thing. Bookmarks might do, too.

My suggestion is this.

We currently have a storage client, which we then exchange for a collection client via `clientForCollection(collection:encrypter:)`. We can then use that collection client for so long as the encrypter is valid — typically the duration of a sync.

Similarly, we should ask our collection client for a batch client, and use that for uploads.

A batch client is created by asking the server for a batch token (presumably also taking the configuration blob as input, when we get around to building that). You can then `.post` records to the batch client, and it'll split them up into chunks for upload, failing if any fail. The caller then commits the batch (optionally doing so in the last chunk) at the end of the appropriate scope.

This has some nice properties:

* You don't need all the records up-front; just use the same batch client each time.
* You can have multiple batches going in parallel, should you be that crazy.
* You can test the batching independently by mocking the collection client itself.
* If you make `Sync15CollectionClient` and the batching version both implement the same interface, you can support batch and non-batch modes by having the batch-me version just return itself, and ignore `commit` calls.

Defining an interface makes you think about the character of the problem. You don't need a new `batchPost` API and a conditional switch between the two; you always `.post`, and then always `.commit` when you're done, and the client you're using does the right thing.

Make sense?

sleroux commented:
Couple of points for clarity:

> Similarly, we should ask our collection client for a batch client, and use that for uploads.

Do you see `storageClient` requesting a `Sync15CollectionClient` vs `Sync15BatchCollectionClient` or internally refactor the existing `Sync15CollectionClient` to use a batch or single client for posting?

> Defining an interface makes you think about the character of the problem. 

Regarding the first question, would this interface lift the existing methods of `Sync15CollectionClient` into a new protocol that is implemented by single and batch variants or do you see this interface as a way for the existing to `Sync15CollectionClient` to proxy it's post requests to the correct implementation internally?

The benefits and every time else makes a lot of sense and sounds like a much better approach to the problem. 

As an aside, I really like hearing and discussing approaches to writing code for the existing sync codebase. I'd like to try keep the code as consistent with the existing model of implementation as possible. In the future, would it be better to document the changes in English instead of code to collaborate on architecture prior to coding?

rnewman commented:
> Do you see `storageClient` requesting a `Sync15CollectionClient` vs `Sync15BatchCollectionClient` or internally refactor the existing `Sync15CollectionClient` to use a batch or single client for posting?

Neither: a batch is a modal operation.

This is a lot like a database: you find the database module (storage client), open a particular database (collection client), start a transaction (batch client), then commit the transaction (commit the batch and throw away the client).

Right now we get a storage client, which controls access to the whole of `/storage/`, use a collection key to get a collection client from it, and then freely `POST` and `GET` that collection. That's like using a database without an explicit transaction.

I think we should continue to do mostly that, with one change: don't `POST` directly with that client. Instead, something like this:

```
  collectionClient.beginBatch() >>== { batch in
    return uploadDeletedRecords(batch)
     >>== uploadModifiedRecords(batch)
     >>==  batch.commit
  }
```

with `uploadDeletedRecords` and friends acting just as they do now:

```
func uploadDeletedRecords(client: POSTableClient15OrWhatever) -> Deferred<…> {
    return self.storage.getDeleted(…) >>== client.uploadRecords
}
```
           
> Regarding the first question, would this interface lift the existing methods…

Both, probably. I expect there'd be a shared interface and some code reuse, perhaps even composition.

> In the future, would it be better to document the changes in English instead of code to collaborate on architecture prior to coding?

Whatever works for you.  Sometimes one needs both :)

sleroux commented:
@rnewman 

Pushed up a commit with my first take on the revised transactional-like API discussed in this thread. 

1. Removed `batchPost` call from the CollectionClient
2. Added new `beginBatch` call to CollectionClient which returns an instance of a `Sync15BatchClient`
3. `Sync15BatchClient` has API for committing the transaction (`commit`) and a call that wraps the posting of records within this batch (`uploadRecords`).
4. Factored out shared post-ing logic from CollectionClient into `POSTClient15` which is passed into `Sync15BatchClient` to reduce duplication of post logic/dependency injection for unit testing.

I've also started to update the synchronizers to use the new batching API. For the most part the API seems to work pretty well but I have a question about the use of the `onUpload` callbacks that get sent in. It looks like the current use of this callback is to advance the timestamp in the mirrorer for the `BufferingBookmarksSynchronizer` and to indicate locally when records have been marked for deletion/synchronized in the case of logins. Is it correct to say,

> The semantics of a batch post should match a single post. That is, with regards to timestamp, we shouldn't advance the timestamp or indicate records have been deleted/synchronized until _after_ the batch commit?

If so, passing in the `onUpload` callback still makes sense but just needs to be called post-commit on the batch instead of each upload. Unfortunately this muddles the meaning on `onUpload` from a consumer standpoint. Maybe a rename makes sense? 

Also, if this is all off base, feel free to let me know!

rnewman commented on Sync/StorageClient.swift:
```diff
@@ -605,6 +622,80 @@ public class Sync15StorageClient {
+    }
+
+    func uploadRecords(lines: [String], ifUnmodifiedSince: Timestamp?) -> Deferred<Maybe<StorageResponse<POSTResult>>> {
+        let batchURI = collectionURI.URLByAppendingPathComponent("batch=\(batchToken)")
```
Is `URLByAppendingPathComponent` correct? It's supposed to be a query parameter… and `UBAPC` encodes its argument.

rnewman replied:
Also, compute the batch collection URI once during init.

sleroux replied:
Yup this was completely wrong. Should be using the `withQueryParams` extension you added to `NSURLExtensions`.

rnewman commented on Sync/StorageClient.swift:
```diff
@@ -605,6 +622,80 @@ public class Sync15StorageClient {
+
+    func commit() -> Deferred<Maybe<StorageResponse<POSTResult>>> {
+        let commitURI = collectionURI.URLByAppendingPathComponent("batch=\(batchToken)").URLByAppendingPathComponent("commit=true")
+        return postClient.post(commitURI, records: [], ifUnmodifiedSince: nil)
```
I expect there will be a significant optimization available on the server if a request is committed immediately, so it behooves us to have a flow that either:

* Doesn't start a batch at all
* Commits immediately.

Give this some thought…

sleroux replied:
I wonder, would it be better to defer the individual batch posting until the commit is called in order to determine if we need to optimize? For example, `uploadRecords` in the batch client queues up pending requests until `commit` is called. At the point of `commit`, we can them add the query param to the last pending request and look for optimizations like doesn't start batch or commits immediately.

rnewman commented on Sync/StorageClient.swift:
```diff
@@ -605,6 +622,80 @@ public class Sync15StorageClient {
+    }
+
+    public func post(uri: NSURL, lines: [String], ifUnmodifiedSince: Timestamp?) -> Deferred<Maybe<StorageResponse<POSTResult>>> {
+        let deferred = Deferred<Maybe<StorageResponse<POSTResult>>>(defaultQueue: client.resultQueue)
```
This class has some smells: it wraps something and is intimately tied to that thing's internals.

Can we effectively fold `POSTClient15` back into `Sync15CollectionClient`?

That is, `Sync15BatchClient` would be obtained from a collection client, and would wrap that collection client. All it would do is pass in the batch token and call commit when done; the regular collection client would still do the record encoding, error handling, etc. Something like:

```
public class Sync15CollectionClient {
    public func postToURI(uri: NSURL, records: …) -> … {
    }
    public func post(records: …) -> … {
        self.postToURI(self.collectionURI, records: records)
    }
    …
    public func beginBatch() -> Deferred<Result<Sync15BatchClient<T>>> {
        self.getBatchToken() >== { token in
            return Sync15BatchClient(token: token, post: self.postToURI)
        }
    }
}
```

Thoughts?

rnewman replied:
I'm not sure if we're able, thanks to generic protocol limitations, but it would be nice if both `Sync15CollectionClient` and `Sync15BatchClient` implemented the same protocol.

sleroux replied:
Yup I believe we can by using `associatedtype` and having the generic passed into each implementor (Batch/Collection clients) fill in the associated type for the prototype. I'll give it a shot!

sleroux replied:
Maybe I'm missing something but it doesn't seem that the collection client and batch client have much in common except for posting records to the server. Even then, the batch client doesn't implement it's own way of posting - it defers to the provided function.

Do you see hiding the batching mechanism all together from consumers (like the synchronizers) to not make modifications to the existing code and infer when a commit/batch is finished?

rnewman commented:
Your understanding of `onUpload` is correct — it should be called when the records have reached the upstream collection. `onStored` would be better, except we tend to use that word for local storage. `onCommit` might be better.

Note that a batch API might thus differ from the non-batch one: for the batch API you would supply the `onCommit` callback *once*, when you start the batch; for as-you-go uploads you'd supply it for each call. So be more invasive if you feel it's right.

rnewman commented on Sync/StorageClient.swift:
```diff
@@ -605,6 +616,210 @@ public class Sync15StorageClient {
+    // For a success response, this will be on the subsequent request, which means we don't
+    // have to worry about handling successes and failures mixed with backoffs here.
+
+    public func commit(ifUnmodifiedSince: Timestamp? = nil, onCollectionUploaded: (POSTResult -> Void)) -> Success {
```
I think IUS should be part of the initializer, no? It's known when the batch begins, and the caller can only call `commit` once.

If multiple entire batches need to be sent (> 10,000 records) then the timestamp will be chained through.

rnewman commented on Sync/StorageClient.swift:
```diff
@@ -605,6 +616,210 @@ public class Sync15StorageClient {
+
+        // Need to deduce how big the data is we're sending across using this reducer
+        let sizeReducer: (total: ByteCount, record: Record<T>) -> ByteCount = { total, record in
+            return total + (self.collectionClient.serializeRecord(record) ?? "").utf8.count
```
Might it not be efficient to serialize to `NSData` at the point of appending?

rnewman replied:
That would be a leaky abstraction, sure.

rnewman commented on Sync/StorageClient.swift:
```diff
@@ -605,6 +616,210 @@ public class Sync15StorageClient {
+
+    public func addRecords(records: [Record<T>]) {
+        log.debug("Adding \(records.count) records into batch")
+        self.records += records
```
If the server supports batching, you should consider chaining out batches as they fill, right here. It'll need a little dancing around a `Deferred`, of course. Follow-up is fine.

rnewman commented on Sync/StorageClient.swift:
```diff
@@ -605,6 +616,210 @@ public class Sync15StorageClient {
+            if records.count > config.maxBatchRecord || sizeOfRecords > config.maxBatchBytes {
+                // TODO: Break into multiple batches
+                return succeed()
+            } else {
```
Early returns.

rnewman commented on Sync/StorageClient.swift:
```diff
@@ -605,6 +616,210 @@ public class Sync15StorageClient {
+
+        // Cut this up into chunks of a maximum size.
+        var batches: [[String]] = []
+        var batch: [String] = []
```
Can't help but feel that there's a more elegant and efficient way to do this…

rnewman commented on Sync/BatchingClient.swift:
```diff
@@ -0,0 +1,263 @@
+
+        // Need to deduce how big the data is we're sending across using this reducer
+        let sizeReducer: (total: ByteCount, record: Record<T>) -> ByteCount = { total, record in
+            return total + (self.serializeRecord(record) ?? "").utf8.count
```
This is really painful. You're serializing every record all the way through to bytes… twice.

rnewman replied:
See comment on line 30.

rnewman commented on Sync/BatchingClient.swift:
```diff
@@ -0,0 +1,263 @@
+public class Sync15BatchClient<T: CleartextPayloadJSON> {
+    private let config: InfoConfiguration
+
+    private var records: [Record<T>] = []
```
You already work with strings as soon as you get to `batchesFromRecords`, and that method immediately serializes. So instead of having this be `[Record<T>]`, I suggest having this be `[(GUID, String)]`; `addRecords` will serialize.

rnewman replied:
Oh, and you probably should keep a running count of bytes as you go.

rnewman commented on Sync/BatchingClient.swift:
```diff
@@ -0,0 +1,263 @@
+
+    public func addRecords(records: [Record<T>]) {
+        log.debug("Adding \(records.count) records into batch")
+        self.records += records
```
Future: if we exceed our batch size, kick off an upload chunk right here. That involves chaining off a `Deferred`.

rnewman replied:
It's worth noting that this isn't really a nice-to-have: the behavior of this batching uploader is noticeably bad in the presence of large numbers of records. Streaming up chunks is the only reasonable way to cap memory usage, and also allows pipelining of record generation and upload.

rnewman commented on Sync/BatchingClient.swift:
```diff
@@ -0,0 +1,263 @@
+            if let batchToken = storageResponse.value.batchToken {
+                log.debug("Uploaded \(lines.count) records and received batch token \(batchToken)")
+                return deferMaybe(batchToken)
+            } else {
```
No need for `else`.

rnewman commented on Sync/BatchingClient.swift:
```diff
@@ -0,0 +1,263 @@
+        // Place the record ID alongside the line text for error reporting later on
+        let idWithLine: (record: Record<T>) -> (GUID, String)? = { record in
+            guard let line = self.serializeRecord(record) else {
+                return nil
```
I'm not confident that this is the best error handling approach. It's probably acceptable for `history`, but totally unacceptable for `bookmarks`. If we can't serialize a record, there's a bug, and we should enter an error handling flow for that (e.g., disable syncing this engine until we upgrade to our next version).

rnewman commented on Sync/BatchingClient.swift:
```diff
@@ -0,0 +1,263 @@
+            let lineSize = line.utf8.count
+
+            if lineSize > Sync15StorageClient.maxRecordSizeBytes {
+                return Maybe(failure: RecordTooLargeError(size: lineSize, guid: id))
```
Do this check waaaaaay earlier: at the very moment you try to add the record to the batcher! I shouldn't get a delayed failure in `.commit` when I can get a synchronous failure at the moment of addition.

rnewman commented on Sync/BatchingClient.swift:
```diff
@@ -0,0 +1,263 @@
+            // Otherwise, add the filled batch to our batches and start a new one.
+
+            // Add 1 to account for the newline
+            guard (size + lineSize + 1) <= config.maxPostBytes && (batch.count + 1) <= config.maxPostRecords else {
```
Someone should check this arithmetic after their first morning coffee! Worth making sure that we'll never produce an empty batch or be unable to enqueue a record that's right at the limit.

rnewman commented on Sync/BatchingClient.swift:
```diff
@@ -0,0 +1,263 @@
+
+        return Maybe(success: batches)
+    }
+}
```
Newline at EOF.

rnewman commented on Sync/BatchingClient.swift:
```diff
@@ -0,0 +1,263 @@
+            // Remove the last batch - we handle the last call in a special case.
+            batch.removeLast()
+
+            // When batching, each upload in the batch is uploaded to a temporary collection until we specif
```
s/specif/specify

rnewman commented on Sync/BatchingClient.swift:
```diff
@@ -0,0 +1,263 @@
+
+                    // Walk through the batches, posting along the way and invoking onUpload
+                    let perChunk: (lines: [String]) -> Success = { lines in
+                        return self.uploader(lines: lines, ifUnmodifiedSince: ifUnmodifiedSince, queryParams: nil)
```
I think this `ifUnmodifiedSince` is wrong. You should use the timestamp from the batch that we just uploaded.

rnewman commented on Sync/BatchingClient.swift:
```diff
@@ -0,0 +1,263 @@
+            var smallerBatch = Batch()
+
+            while size <= config.maxTotalBytes && recordCount <= config.maxTotalRecords && !mutBatch.isEmpty {
+                let partialBatch = mutBatch.removeFirst()
```
All of this nonsense becomes unnecessary if you do the work at the point of accumulation.

That is: when the caller calls `addRecord`, serialize the incoming record. If it fails, stop. If it succeeds, figure out how many bytes it'll be. If it's too big for one record, stop. Otherwise, accumulate it into a new or existing batch, and when a batch is full, accumulate that into a new or existing bunch.

Have tiers like this:

```
typealias UploadRecord = (guid: GUID, payload: String, size: Int)

class Payload {
    let config: InfoConfiguration
    var records: [UploadRecord] = []
    var bytes: Int = 0
    var count: Int = 0

    func take(record: UploadRecord) -> Bool {
        if (record.size + bytes) > config.maxPostBytes {
            return false
        }
        if count >= config.maxPostRecords {
            return false
        }
        …
        return true
}

class Batch {
     …
}

```

and so on. At each level you can keep track of bytes and counts. Note that streaming uploads (a good idea) involves keeping track of request, record, and byte counts so that you can choose when to start a new bunch.

rnewman replied:
Your ideal goal is for an incoming record to be immediately serialized, for a request to be sent as soon as you have enough records, and for a batch to be committed as soon as you run out of records (i.e., `commit` is called) or the batch is full and a new bunch needs to be started.

sleroux replied:
You mentioned something similar in a comment above about firing off a batch upload from `addRecords` when the batch is filled. When you talk about streaming, do you see us performing network requests as we go prior to the commit if our batch is full? I wonder if that has problematic implications from a user-perspective because once we've sent records to the server, we're committing those records to the server before the user calls commit. Maybe the abstraction of a database transaction is incorrect here and we should move more into a streaming model?

rnewman replied:
Yes, I see us sending requests before `commit`, *if the server supports batching* (and perhaps even if it doesn't, if the caller is willing).

The transaction concept still applies. Think about how a database works and its interface. You call `BEGIN TRANSACTION`, and can then do whatever you like for however long you like. At some point you call `ROLLBACK` or `END TRANSACTION`, and you expect the externally visible state of the database to either solidify or return to what it was. That doesn't mean that the database isn't writing stuff to disk during the transaction — indeed, it almost certainly is!

The same applies here. When you have a batch token, it means you can write whatever you want to the server, and it doesn't get *committed* until you make a request with `commit=true`. If `commit` is never called, the batch will be abandoned and will be garbage collected. The sensible thing to do, then, is to start making those HTTP requests as soon as we can — you might be uploading megabytes of records, and it doesn't make much sense to wait until we have everything before doing so.

Perhaps the best way to think about this is in terms of Clojure-esque channels and transducers. A channel is a producer-driven queue, like a conveyor belt. A transducer is a worker between two conveyor belts. In this case, our caller is pushing records onto the conveyor belt. We have a transducer that's bundling up records into collections of 50/1MB, and one that's shifting a conveyor belt into different bunches as a limit is reached (similar process). The consumer is uploading each bundle once it arrives. Finally, you tell the destination to commit when the channel is closed.

If the server doesn't support batching, then yes, it's unsafe to leave large gaps between uploads. But we do already upload in chunks without batching support, and it's likely that more records will be delivered for upload before our first upload is complete, so the only concern would be if we aborted the entire upload operation due to a malformed record or were interrupted (partial write). That's why we built atomic uploads in the first place, so I wouldn't spend too much time worrying about that.

From the caller's perspective, the contract remains: the records it added to the batcher will be permanently committed to the server at some point between calling `commit` and the `onCollectionUploaded` callback being invoked. We just happen to sometimes move them up to the server before `commit` is called.

sleroux replied:
Thanks for the explanation! I completely disregarded the behavior of not calling commit and it's effect of not finishing the batch. This makes a lot of sense.

> Perhaps the best way to think about this is in terms of Clojure-esque channels and transducers. A channel is a producer-driven queue, like a conveyor belt. A transducer is a worker between two conveyor belts. In this case, our caller is pushing records onto the conveyor belt. We have a transducer that's bundling up records into collections of 50/1MB, and one that's shifting a conveyor belt into different bunches as a limit is reached (similar process). The consumer is uploading each bundle once it arrives. Finally, you tell the destination to commit when the channel is closed.

Great analogy by the way 👌

rnewman commented on Sync/Info.swift:
```diff
@@ -50,3 +50,24 @@ public class InfoCollections {
+    // Maximum total combined size in bytes of the record payloads that can be uploaded to a collection
+    // as part of a batched upload
+    public let maxTotalBytes: Int
+}
```
Newline at EOF.

sleroux commented:
@rnewman  And the saga continues!

I've pushed up a commit which (I think) addresses all the comments mentioned. Some highlights:

1. Instead of waiting until the `commit` before talking to the server, records are sent when the batch client has enough to send.

2. Serialization and error checking is handled at the point of adding records instead of waiting until the commit.

3. `Payload` and `BatchMeta` encapsulate limit checking and properties for keeping track of how much data we've inputted into the payload and how much we've sent respectively.

4. Removed explicit non-batch behaviour. When sending records as part of the start, if we don't get back a token, we just did a POST, so we don't need to do anything special. Just invoke the callbacks and move the timestamp and continue onwards.

5. Move the `ifUnmodifiedSince` timestamp forward when committing requests to the server.

I think this cleans up a lot of tangled stateful behaviour I was doing before.

Also, I haven't touched the tests yet so that will most likely fail 😬

rnewman commented:
I'll be looking at this today.

rnewman commented on Sync/BatchingClient.swift:
```diff
@@ -0,0 +1,221 @@
+        return true
+    }
+
+    private func payloadFitsInBatch(payload: Payload) -> Bool {
```
Perhaps just inline this?

rnewman commented on Sync/BatchingClient.swift:
```diff
@@ -0,0 +1,221 @@
+
+    public func addRecord(record: Record<T>) -> Success {
+        // Sanity checks and assumptions
+        precondition(config.maxPostBytes < config.maxTotalBytes)
```
> In `-O` builds (the default for Xcode's Release configuration): if `condition` evaluates to `false`, stop program execution.

You probably want to return failure here instead.

rnewman replied:
That is: if we push a server config that gets this wrong, making iOS browsers silently crash every few minutes is unlikely to be the best choice.

rnewman commented on Sync/BatchingClient.swift:
```diff
@@ -0,0 +1,221 @@
+    public func addRecord(record: Record<T>) -> Success {
+        // Sanity checks and assumptions
+        precondition(config.maxPostBytes < config.maxTotalBytes)
+        precondition(config.maxPostRecords < config.maxTotalRecords)
```
Also I think these should both be `<=`. But think carefully!

rnewman commented on Sync/BatchingClient.swift:
```diff
@@ -0,0 +1,221 @@
+
+                // Based on the precondition, we must be able to add at least one payload into the batch 
+                // without it failing
+                >>> effect({ self.batchMeta.add(self.currentPayload) })
```
`batchMeta` should be checked — if not necessarily incremented — before line 135. If we're on a near-empty non-first batch, and we've uploaded 9,999,999 bytes, `batchMeta.add` will fail even if `currentPayload.add` succeeds. As we process an incoming record we should check *both* of those conditions before taking action.

rnewman replied:
Oh, my bad — this *is* the first-batch branch.

rnewman commented on Sync/BatchingClient.swift:
```diff
@@ -0,0 +1,221 @@
+                // Based on the precondition, we must be able to add at least one payload into the batch 
+                // without it failing
+                >>> effect({ self.batchMeta.add(self.currentPayload) })
+                >>> effect({ self.currentPayload = Payload(config: self.config) })
```
You can combine all of these `effect` blocks, by definition.

sleroux replied:
Also Grisha pointed out on IRC that there is a bug here that the record never gets added to the new payload. I've added a fix for that.

rnewman commented on Sync/BatchingClient.swift:
```diff
@@ -0,0 +1,221 @@
+            return self.start(self.currentPayload)
+                >>== effect({ self.batchToken = $0 })
+                >>> effect({ self.batchMeta = BatchMeta(config: self.config) })
+                >>> effect({ self.batchMeta.add(self.currentPayload) })
```
Similarly: combine.

rnewman commented on Sync/BatchingClient.swift:
```diff
@@ -0,0 +1,221 @@
+        guard let token = batchToken else {
+            return uploadPayload(self.currentPayload)
+                >>== effect({ self.ifUnmodifiedSince = $0.value.modified })
+                >>== effect({ self.onCollectionUploaded?($0.value) })
```
Same.

rnewman commented on Sync/BatchingClient.swift:
```diff
@@ -0,0 +1,221 @@
+            log.debug("Uploaded \(lines.count) records but received no batch token - records will be committed to collection right away")
+            self.ifUnmodifiedSince = storageResponse.value.modified
+            self.onCollectionUploaded?(storageResponse.value)
+            return deferMaybe(nil)
```
`return succeed()`

rnewman commented on Sync/BatchingClient.swift:
```diff
@@ -0,0 +1,221 @@
+
+        // We could send a payload when committing the batch but that would involve us slicing up the in-progress 
+        // payload to make it fit just right. Easier to just push it up in a new batch.
+        return self.uploadPayload(Payload(config: self.config), queryParams: [batchParam, commitParam])
```
I'm confused. `addRecord` never allows the `Payload` to expand beyond its allowable size. So we never need to slice up a payload; we can always upload it here.

Furthermore, I don't see how an upload happens in the simplest case!

```
let client = Sync15BatchClient<T>(…)
client.addRecord(myRecord)
// 50 times.
client.addRecord(anotherRecord)   // 51st, so there's a token
client.endBatch()
```

… looks like that adds the last record to the new empty payload, calls `commit`, which then sends an empty payload, leaving the final record not uploaded.

sleroux replied:
Ah yup. Looks like I was getting a bit mixed up with payloads vs records.

This should just send the current payload instead of an empty one.

rnewman commented on Sync/StorageClient.swift:
```diff
@@ -605,6 +617,12 @@ public class Sync15StorageClient {
+                                                         maxPostRecords: 100,
+                                                         maxPostBytes: 1048576,
+                                                         maxTotalRecords: 10000,
+                                                         maxTotalBytes: 104857600)
```
Insert Swifty `_`:

```
     maxTotalBytes: 104_857_600
```

rnewman commented on SyncTests/BatchingClientTests.swift:
```diff
@@ -0,0 +1,257 @@
+        let onCollectionUpload: (POSTResult -> Void) = { _ in collectionUploadCount += 1 }
+
+        let batch = Sync15BatchClient(config: miniConfig, ifUnmodifiedSince: nil, serializeRecord: serializeRecord, uploader: uploader)
+        let records = generateMockRecords(miniConfig.maxPostRecords)
```
White-box testing! `maxPostRecords + 1` will verify whether my observation above is actually a bug.

rnewman commented on SyncTests/BatchingClientTests.swift:
```diff
@@ -0,0 +1,257 @@
+
+    private func mockRecord() -> Record<CleartextPayloadJSON> {
+        let jA = "{\"id\":\"mock\",\"histUri\":\"http://foo.com/\",\"title\": \"ñ\",\"visits\":[{\"date\":1222222222222222,\"type\":1}]}"
+        return Record<CleartextPayloadJSON>(id: "mock",payload: CleartextPayloadJSON(JSON.parse(jA)), modified: 10000, sortindex: 123, ttl: 1000000)
```
Spaces and newlines, please :)

rnewman commented on SyncTests/BatchingClientTests.swift:
```diff
@@ -0,0 +1,257 @@
+        batch.commit(onCollectionUpload).succeeded()
+
+        // Should only have called the upload/collection callbacks once
+        XCTAssertEqual(collectionUploadCount, 1)
```
You need to check how many records were uploaded, too!

rnewman commented on SyncTests/BatchingClientTests.swift:
```diff
@@ -0,0 +1,257 @@
+        var collectionUploadCount = 0
+        var uploadOpCount = 0
+
+        let uploader: BatchUploadFunction = { lines, timestamp, queryParams in
```
An unused variable here is a warning sign. The very most important thing this uploader needs to do is make sure that all the records make it to the server, so take a look at `lines`.

rnewman commented on SyncTests/BatchingClientTests.swift:
```diff
@@ -0,0 +1,257 @@
+        batch.commit(onCollectionUpload).succeeded()
+
+        // Should only have called the upload/collection callbacks once
+        XCTAssertEqual(collectionUploadCount, 1)
```
Same. Check that we have the right number of records. (Extra points for being thorough and making sure they're the right records, and not just the same payload over and over!)

rnewman commented on Sync/BatchingClient.swift:
```diff
@@ -0,0 +1,221 @@
+    private var currentPayload: Payload
+    private var onCollectionUploaded: (POSTResult -> Void)?
+
+    // TODO: use I-U-S.
```
Is this still a TODO?

rnewman commented on Sync/BatchingClient.swift:
```diff
@@ -0,0 +1,221 @@
+    private var onCollectionUploaded: (POSTResult -> Void)?
+
+    // TODO: use I-U-S.
+    // Each time we do the storage operation, we might receive a backoff notification.
```
This comment needs to be updated, and we need to make sure if our backoff handling is correct.

What happens if we get an `X-Backoff` header in response to one of our payloads? Do we give up or blindly continue? How do we make that decision?

sleroux replied:
Right now the backoff check is done in the `post` call from the collection client that gets passed in as the uploader function. The post call already handles filling a deferred with the `ServerInBackoffError` error. Tracing the `uploader` calls in the `BatchingClient` shows that this error will propagate up to the `addRecord` and `endBatch`. Since the `addRecord` would return a failure, external clients and bubble this up further following the same logic flow the existing `post` call has. I don't think any changes need to be done here.

rnewman commented on Sync/BatchingClient.swift:
```diff
@@ -0,0 +1,221 @@
+    // For a success response, this will be on the subsequent request, which means we don't
+    // have to worry about handling successes and failures mixed with backoffs here.
+
+    init(config: InfoConfiguration, ifUnmodifiedSince: Timestamp? = nil, serializeRecord: (Record<T>) -> String?, uploader: BatchUploadFunction, onCollectionUploaded: (POSTResult -> Void)?) {
```
Sidenote: why is `onCollectionUploaded` optional?

sleroux replied:
Probably doesn't need to be. I always thought of the onCollectionUploaded being an optional callback but we're always going to be passing something in from the consuming side so it might as well be non-optional.

rnewman commented:
This is looking really good. One possible bug, some additional necessary test coverage, and some questions.

sleroux commented:
Related to the headers being sent back and forth,

1. Should I also include the `X-Weave-*` headers in our requests from the batch client? This documentation [1] shows sending this information along with the request to give the server a heads up. Shouldn't be difficult to add since we already have those calculations in the client.

2. My current assumption with how `X-If-Unmodified-Since` works is that a collection is unmodified until either a batch commits or in the case of unsupported batching, the post completes. The timestamp shouldn't move forward for payloads requests sent during a batch, correct?

[1] https://github.com/mozilla-services/docs/pull/60/files#diff-f817ccb87b7caca4fb26e0f4dc2036aeR524

sleroux commented:
> My current assumption with how X-If-Unmodified-Since works is that a collection is unmodified until either a batch commits or in the case of unsupported batching, the post completes. The timestamp shouldn't move forward for payloads requests sent during a batch, correct?

I'm also getting confused by the difference between this and the usage of `lastModified` in the code. I'm looking at using the new batch client in the `IndependentRecordSynchronizer` which passes along the lastModified value through the chain and returns a `DeferredTimestamp`.

rnewman commented:
We will know `X-Weave-Bytes`, assuming I'm reading the docs correctly, but in the general case we don't know those values at the time that we send the `batch=true` request. In the case where `commit` has been called before the first request, we can include them. You'll need to ask for clarification about whether it's possible or worthwhile to include e.g., `X-Weave-Total-Bytes` in a non-first request.

We ought to double check that we send the correct value for each, because the server docs I believe use:

* "payload" = a record
* "upload" = a POST
* "batch" = a collection of uploads followed by a commit
* no name for a bundle.

rnewman commented:
Re `X-I-U-S`: we should include the same `X-I-U-S` on every request within a batch. If concurrent modification to the collection occurs on batch creation, update, or commit, then the batch will fail and we'll get a `412` response.

We can sanity-check the `X-Last-Modified` response header during the batch (already exposed by our Swift storage API), which should match our `X-I-U-S`.

This is why we chain `lastModified`: it's the modified timestamp for the collection (as opposed to `X-Weave-Timestamp`, which is the server timestamp at the point of handling the request — usually the same thing, but not precise). If you want an uninterrupted sequence of uploads, you use the `lastModified` value from the previous success as the `ifUnmodifiedSince` condition of the next.

Within a batch `lastModified` won't change unless there's a race.

Note that for a collection sync as a whole, the `lastModified` value actually comes at the end of the previous `GET`!

Note also that `X-I-U-S` behaves differently for `POST` (collection timestamp) and `PUT` (per-record timestamp). That's only relevant if you start poking at the `clients` engine.

sleroux commented:
Pushed up some changes for handling timestamps and started to replace the `LoginSynchronizer`/`IndependentRecordSynchronizer` code to use the batching client. Couple of questions:

1. I've replaced the `uploadRecordsinChunks` method to use the batching client but haven't touched the `uploadRecords(by)` code. Looks like the `by` parameter is used to control the number of records we send per a post but now with batching is this still needed?  I can see us keeping it around to have control over the flow of records to the server instead of always trying to fill payloads to the max but not sure if its worth it.

2. I'm trying to connect to the staging environment to test out the batching API but none of the existing `FirefoxAccountConfiguration`s seem to work. When I try using the `StableDev` configuration which uses staging token server, I get back unauthorized from the token server. I feel like I'm missing something here...

rnewman commented:
`uploadRecords(by)` isn't needed once all consumers use the new batching uploader. I think we'd control the flow of records to the server by passing a different record limit to the new uploader.

You should ask @rfk or bobm in `#sync` if they can point you to the right place for a working staging config. It might have changed since that stuff was committed to our repo. If so, commit a fix!

rnewman commented on Sync/BatchingClient.swift:
```diff
@@ -0,0 +1,230 @@
+
+/**
+ *  A payload represents a block of records sent in a single POST to the server within a batch.
+ *  The payload tries to contains as many records it can fit while respecting the limits imposed
```
s/contains/contain

rnewman commented on Sync/BatchingClient.swift:
```diff
@@ -0,0 +1,230 @@
+    private var currentPayload: Payload
+    private var onCollectionUploaded: (POSTResult -> DeferredTimestamp)
+
+    // TODO: use I-U-S.
```
Fix comment.

rnewman commented on Sync/BatchingClient.swift:
```diff
@@ -0,0 +1,230 @@
+    }
+
+    public func addRecords(records: [Record<T>]) -> Success {
+        return walk(records, f: addRecord)
```
This is elegant, but the initial serialization and construction of a batch are all synchronous, so we can be more efficient than this.

Make `serializeRecord` throw, `map` to serialize lines, `throw` if a record is too big; wrap that all in a `catch` that rejects.

Now you have a sequence of `UploadRecord`s. Sequentially accumulate these in `currentPayload` until you get `false`. In the best case — a small number of records — we create only one `Deferred` and do no extra work.

rnewman replied:
Specifically, I'd like to see `addRecords` itself look like a synchronous block that consumes all possible records from the sequence, followed by either `return succeed()` or an upload chained into a recursive call to `addRecords`. Even something like this (you don't have to agree, but this is illustrative):

```
public func addRecords(records: [Record<T>]) -> Success {
  if (records.isEmpty) {
    return succeed()
  }
  return self.addRecords(records.makeIterator())
}

public func addRecords(records: Iterator<Record<T>>) -> Success {
  while let record = records.next() {
    if let def = self.accum(record) {
      return def >>> { self.addRecords(records) }
    }
  }
  return succeed()
}

// Throws on failure. 
// Returns a Deferred iff we have started an upload *and* added the provided
// record to be uploaded.
// Returns nil if the record was successfully added to the
// current payload but no upload was triggered.
private func accum(record: Record<T>) -> Success? throws {
  guard let line = … {
    …
    return nil
  }
  …
}
```

sleroux replied:
I like how this breaks down the operations logically into something that more resembles your conveyor belt analogy you used before. Pull off some records, bunch them up, add them while respecting the state of uploading. This would also clean up the current `addRecord` call but separating the operations a bit. I'll play around with this approach and see what I can do along with the other comments in the PR about atomically swapping out the payload addition.

rnewman replied:
Aye. The two main thrusts of my comments are to make sure we don't accidentally mishandle a payload, and to make the simple case efficient. 👍

rnewman commented on Sync/BatchingClient.swift:
```diff
@@ -0,0 +1,230 @@
+        // Otherwise, attempt to start a new batch. If we fail starting - no problem! The records would
+        // have already been sent in a single POST call.
+        guard let token = batchToken else {
+            return self.start(self.currentPayload)
```
`start` should 'steal' `currentPayload` and (pseudo-atomically) replace `self.currentPayload` with a new empty payload. You don't want two `addRecord` calls to end up with `self.start` being called twice with the same payload…

rnewman commented on Sync/BatchingClient.swift:
```diff
@@ -0,0 +1,230 @@
+                // without it failing
+                >>> effect({
+                    self.batchMeta.add(self.currentPayload)
+                    self.currentPayload = self.addRecordToNewPayload(uploadRecord)
```
… and if you split out the payload handling like that, then here you can just chain back into `addRecord` (or, better, `addRecords`!).

rnewman commented on Sync/Synchronizers/LoginsSynchronizer.swift:
```diff
@@ -107,20 +108,22 @@ public class LoginsSynchronizer: IndependentRecordSynchronizer, Synchronizer {
+        return uploadRecordsInChunks(records,
+                                     lastTimestamp: lastTimestamp,
+                                     storageClient: storageClient,
+                                     onUpload: onUpload) >>> effect({ log.debug("Done syncing.") })
```
Remove debug statement.

rnewman commented on Sync/Synchronizers/LoginsSynchronizer.swift:
```diff
@@ -107,20 +108,22 @@ public class LoginsSynchronizer: IndependentRecordSynchronizer, Synchronizer {
-    private func uploadDeletedLogins(guids: [GUID], lastTimestamp: Timestamp, fromStorage storage: SyncableLogins, withServer storageClient: Sync15CollectionClient<LoginPayload>) -> DeferredTimestamp {
+            // Figure out which ones the server deleted and which ones were modified
+            let guidsDeleted = deleted.filter(uploaded.contains)
+            let guidsModified = modified.filter(uploaded.contains)
```
Consider using sets here; set intersection should be faster, and we don't care about ordering.

rnewman commented on SyncTests/BatchingClientTests.swift:
```diff
@@ -0,0 +1,383 @@
+    }
+
+    func testFailToSerializeRecord() {
+        let uploader: BatchUploadFunction = { _ in deferEmptyResponse() }
```
Worth defining an overloaded version of `deferEmptyResponse`?

rnewman commented on Sync/BatchingClient.swift:
```diff
@@ -0,0 +1,230 @@
+        if batchMeta.add(self.currentPayload) {
+            return self.uploadPayload(self.currentPayload, queryParams: [NSURLQueryItem(name: "batch", value: token)])
+                >>> effect({
+                    self.currentPayload = self.addRecordToNewPayload(uploadRecord)
```
This really confirms my feeling that one spot should be responsible for atomically swapping out a new payload. There should be only two ways to get a new payload:

- When an upload begins.
- When a batch boundary is hit.

sleroux commented:
@rnewman 

Alright I think I got a better version of this for you to look at.

* I refactored the BatchingClient to better support adding multiple records by making the flow more synchronous in nature. 
* Fixes various bugs that were happening at edge cases. For example, when batching is supported and a single payload was being sent, a batch was started by not completed. In the case of a single upload, batch=true&commit=true are sent.
* Added more test cases to cover said edge cases.
* Removed the modified property from `POSTResult` in favor of using `X-Last-Modified` instead. Now the `ifUnmodifiedSince` timestamp moves forward using that instead of the old JSON property.
* Added a working staging configuration for testing against the actual API. From what I've seen in my tests with syncing bookmarks/logins, it seems to work pretty well 👍
* Address other comments such as using Sets for better performance when finding modified/delete records, comment issues, etc

One important change I made that's worth mentioning is now the `onCollectionUpload` callback is called for every batch start and commit instead of just the commit. The start part was missing before which I think is incorrect since we want to update our time stamps with what we received at the start of the batch then move forward when the batch completes.

Lastly, I've included a new Sync scheme so you can run just the Sync unit tests with.

rnewman commented on Sync/BatchingClient.swift:
```diff
@@ -25,14 +25,20 @@ private let log = Logger.syncLogger
     let config: InfoConfiguration
+
+    var isEmpty: Bool {
+        return records.isEmpty
```
`count == 0` matches `Batch`.

rnewman commented on Sync/StorageClient.swift:
```diff
@@ -232,18 +232,16 @@ public struct StorageResponse<T> {
 }
 
-public typealias BatchToken = String
+public typealias BatchToken = Int64
```
The docs say this is opaque:

```
 "batch": "OPAQUEBATCHID",
```
so this should be a string.

rnewman commented on Sync/BatchingClient.swift:
```diff
@@ -0,0 +1,245 @@
+    // Batch is full if we can't add another payload to it
+    var isFull: Bool {
+        return (config.maxPostRecords + count > config.maxTotalRecords) ||
+               (config.maxPostBytes + bytes > config.maxTotalBytes)
```
This might warrant an explanatory comment, because I had to stare at it for a minute to figure out why you were comparing across 'domains'.

You're saying here that if:

- The number of records we've already sent, plus another full POST, exceeds the total the server will accept, or
- The number of bytes we've already sent, plus another full POST, exceeds the total the server will accept, then
- We can't start another payload in this batch.

Chewing on that for a minute: I think that's safe, but not really optimal. What if the next POST will only include one small record? We'll instead say this batch is full and commit it, moving on to another batch in the bundle, right?

sleroux replied:
> 
>     The number of records we've already sent, plus another full POST, exceeds the total the server will accept, or
>     The number of bytes we've already sent, plus another full POST, exceeds the total the server will accept, then
>     We can't start another payload in this batch.

Correct.

> What if the next POST will only include one small record? We'll instead say this batch is full and commit it, moving on to another batch in the bundle, right?

Yup that's what would happen. I spoke with Grisha yesterday and he's doing the check at the record/byte level so we're not sending small payloads if they would have fit in the existing batch. This can definitely be made smarter.

rnewman commented on Sync/BatchingClient.swift:
```diff
@@ -0,0 +1,245 @@
+
+    private var bytes: ByteCount = 0
+    private var count: Int = 0
+    private var token: BatchToken
```
Move all of these simple `var`s to the top of the `struct` definition.

rnewman replied:
And shouldn't `token` be a `let`?

sleroux replied:
Yup. Must have been leftover from some in progress work.

rnewman commented on Sync/StorageClient.swift:
```diff
@@ -232,18 +232,29 @@ public struct StorageResponse<T> {
     }
 }
 
+public typealias BatchToken = Int64
```
As noted on an earlier diff: the batch token is an opaque identifier, and is delivered as a JSON string, according to the docs.

rnewman commented on Sync/BatchingClient.swift:
```diff
@@ -0,0 +1,245 @@
+        func uploadPayload() -> Success {
+            let lines = self.currentPayload.records.map { $0.payload }
+            let batchStartParam = NSURLQueryItem(name: "batch", value: "true")
+            let commitParam = NSURLQueryItem(name: "commit", value: "true")
```
Maybe pull these out as `static` or `private let`?

sleroux replied:
That would be much better. I think I use them in other places too...

rnewman commented on Sync/BatchingClient.swift:
```diff
@@ -0,0 +1,245 @@
+                if let uploadOp = accumulateRecord(serializedRecord) {
+                    // Since we couldn't add this record, throw it onto the rest and try adding everything again.
+                    let restOfRecords = generator.map { $0 }
+                    return uploadOp >>> { self.addRecords([record] + restOfRecords) }
```
Rather than consuming the entire generator, concatenating a whole new array, and starting again, do something like this:

```
if let uploadOp = accumulateRecord(serializedRecord) {
    // An upload started. Chain off it to add the triggering record, and pick up
    // where we left off.
    return uploadOp >>> {
        let deferred = accumulateRecord(serializedRecord) ?? succeed()  // Might need type fiddling.
        return deferred >>> self.addRecords(generator)     // Add the generator variant.
    }
```

rnewman replied:
Note that because we only ever recurse with the generator, we're sure that this will terminate!

rnewman commented on Sync/BatchingClient.swift:
```diff
@@ -0,0 +1,245 @@
+        // Push up the current payload to the server and reset
+        let lines = self.currentPayload.records.map { $0.payload }
+        return uploader(lines: lines, ifUnmodifiedSince: self.ifUnmodifiedSince, queryParams: batch.batchParams)
+            >>== effect({ _  in self.currentPayload = Payload(config: self.config) })
```
Is there a good reason for delaying reassigning `currentPayload` until the upload succeeds? ISTM that you could do this right after computing `lines`, save yourself the memory burden, and avoid the possibility of accidentally adding records to a payload that's already being uploaded.

sleroux replied:
Nope no reason. I'll move it before the uploading instead of after.

rnewman commented on Sync/BatchingClient.swift:
```diff
@@ -0,0 +1,245 @@
+            >>== effect(moveForward)
+            >>== effect({  _ in
+                self.currentBatch = nil
+                self.currentPayload = Payload(config: self.config)
```
Same here.

rnewman replied:
Indeed, one might:

```
private func freeze() -> [String] {
    let lines = self.currentPayload.records.map { $0.payload }
    self.currentPayload = Payload(config: self.config)
    return lines
}
```

then do

```
let lines = self.freezePayload()
self.currentBatch = nil
return uploader(lines: lines, …) >>== effect(moveForward)
```

sleroux replied:
Ohhh I like `freeze`

rnewman commented on Sync/BatchingClient.swift:
```diff
@@ -0,0 +1,245 @@
+                    self.currentBatch = startedBatch
+                }
+
+                self.currentPayload = Payload(config: self.config)
```
Same here.

rnewman commented on Sync/BatchingClient.swift:
```diff
@@ -0,0 +1,245 @@
+             >>== { response in
+                if let token = response.value.batchToken {
+                    var startedBatch = Batch(config: self.config, token: token)
+                    startedBatch.add(self.currentPayload)
```
(This would need capturing, certainly)

rnewman commented on Sync/Synchronizers/Bookmarks/BookmarksSynchronizer.swift:
```diff
@@ -59,23 +59,22 @@ class TrivialBookmarkStorer: BookmarkStorer {
-            local.setModifiedTime(modified, guids: result.success)
-            return deferMaybe(result.modified)
+            log.debug("Uploaded records got timestamp \(lastModified).")
+            local.setModifiedTime(lastModified ?? modified, guids: result.success)
```
AFAICS `modified` is no longer set, so it will always be `0`. Is that what you want?

sleroux replied:
Nope. I guess the annoying thing was that `modified` was non-optional before but now because its coming down as a header value its optional. What are your thoughts on force unwrapping the XLM timestamp? Doing this would match the previous behavior of always having a timestamp but introduce the unfortunate effect of a crash if it's not there. There isn't really an empty state for a timestamp.

sleroux replied:
I guess we're using 0 as a default timestamp in case the `X-Weave-Timestamp` property isn't in the request. We can always do that for the `XLM` header.

rnewman commented on Sync/BatchingClient.swift:
```diff
@@ -0,0 +1,245 @@
+        do {
+            var generator = records.generate()
+            while let record = generator.next() {
+                let serializedRecord = try serialize(record)
```
It's important to aggressively serialize outgoing records.

If you take a look at `IndependentRecordSerializer`, you'll see that if any record we plan to upload is over-sized, we upload *nothing*.

That's not the case with your refactoring here, if the server doesn't support batching (which will be the case for a while): we will accumulate the first fifty records and start an upload *without trying to serialize the 51st*. If a later record is too large, we will give up, leaving a partial write on the server.

So rework `addRecords` to be like this:

```
public func addRecords(records: [Record<T>]) -> Success {
    if records.isEmpty {
        return succeed()
    }
    do {
        let serialized = try records.map { try serialize($0) }     // or however Swift likes it
        return addRecords(serialized.generate())
    } catch let e {
        return deferMaybe(e as! MaybeErrorType)
    }
}

public func addRecords(records: Generator<UploadRecord>) -> Success {
    …
}
```

rnewman commented on Sync/Synchronizers/LoginsSynchronizer.swift:
```diff
@@ -107,20 +108,20 @@ public class LoginsSynchronizer: IndependentRecordSynchronizer, Synchronizer {
-            storage.markAsDeleted($0.success) >>> always($0.modified)
+        let onUpload: (POSTResult, Timestamp?) -> DeferredTimestamp = { result, lastModified in
+            let uploaded = Set(result.success)
+            let deletedGUIDs = Array(uploaded.intersect(deleted))
```
Can you just make `markAsDeleted` and `markAsSynchronized` take a `SequenceType` or other abstraction and avoid the `Array` alloc? All it needs is to be able to walk the sequence…

sleroux commented:
Added some more commits to fix up the review comments and a slight rewriting of `accumulateRecord` to fit records better into batches. I ended up scraping the Batch/Payload structs and replaced them with counts that match more closely to the info/configuration endpoint (postBytes/postCount/totalBytes/totalCount). The payload abstraction was getting in the way of fitting records better and I think this cleans it up even further.

sleroux commented:
Also shout out to my unit tests for letting me constantly refactor this stuff 💯

rnewman commented on Sync/BatchingClient.swift:
```diff
@@ -0,0 +1,218 @@
+
+private let log = Logger.syncLogger
+
+private typealias UploadRecord = (guid: GUID, payload: String, size: Int)
```
Nit: `size` -> `sizeBytes`

rnewman commented on Sync/BatchingClient.swift:
```diff
@@ -0,0 +1,218 @@
+    private let uploader: BatchUploadFunction
+    private let serializeRecord: (Record<T>) -> String?
+
+    private let commitParam = NSURLQueryItem(name: "commit", value: "true")
```
Lift out of this class.

rnewman commented on Sync/BatchingClient.swift:
```diff
@@ -0,0 +1,218 @@
+
+        if let token = self.batchToken {
+            return commitBatch(token) >>> succeed
+        } else {
```
No need for `else` after `return`.

rnewman commented on Sync/BatchingClient.swift:
```diff
@@ -0,0 +1,218 @@
+            let lines = self.freezePost()
+            let queryParams = [
+                batchQueryParamWithValue("true"),
+                NSURLQueryItem(name: "commit", value: "true")
```
Use `commitParam`!

rnewman replied:
But I'm not sure what you're doing here. We're not currently in a batch, because `batchToken` is `nil`. We've been told to end a batch, but there isn't one. Here you're sending query parameters that imply both starting and ending a batch. Why bother? Just send the request with no params!

rnewman commented on Sync/BatchingClient.swift:
```diff
@@ -0,0 +1,218 @@
+        }
+
+        do {
+            let serialized = try records.map { try serialize($0) }
```
Worth adding a comment to the effect of some of my review comments: we eagerly serialize all records before we queue them, because if we're going to fail we want to do it early.

rnewman commented on Sync/BatchingClient.swift:
```diff
@@ -0,0 +1,218 @@
+        }
+    }
+
+    private func addRecords(generator: IndexingGenerator<[UploadRecord]>) -> Success {
```
```
addRecords(var generator: …) …
```

means you don't need `mutGenerator`.

sleroux replied:
Mutable parameters are deprecated in Swift 3 so a warning shows when I do this :(

rnewman commented on Sync/BatchingClient.swift:
```diff
@@ -0,0 +1,218 @@
+        while let record = mutGenerator.next() {
+            if let uploadOp = accumulateRecord(record) {
+                return uploadOp >>> {
+                    if let deferred = self.accumulateRecord(record) {
```
Why the two levels of nesting? Surely

```
while let record = generator.next() {
    if let upload = accumulateRecord(record) {
        return uploadOp >>> { self.addRecords(generator) }
    }
}
return succeed()
```

is equivalent?

rnewman commented on Sync/BatchingClient.swift:
```diff
@@ -0,0 +1,218 @@
+                    addToBatch(record)
+                    return nil
+                } else {
+                    // Otherwise, we don't added the record and just POST away.
```
We didn't add the record — it fits in this batch, but the POST would have been too big, so `addToPost` returned `false`. Then we get to here and upload the current POST with `postInBatch`. Don't we lose `record` here?

rnewman replied:
Oh, is this the source of that confusing double-attempt above? If so, please tidy that up (perhaps moving it here) or comment `accumulateRecord` to explain!

rnewman commented on Sync/BatchingClient.swift:
```diff
@@ -0,0 +1,218 @@
+        let lines = self.freezePost()
+        let queryParams = [
+            batchQueryParamWithValue(token),
+            NSURLQueryItem(name: "commit", value: "true")
```
Use the extracted commit param.

rnewman commented on Sync/BatchingClient.swift:
```diff
@@ -0,0 +1,218 @@
+             >>== { response in
+                if let token = response.value.batchToken {
+                    self.batchToken = token
+                    self.totalRecords = postRecordCount
```
Shouldn't these be `+=`?

sleroux replied:
Nope because since we're starting a batch, the assumption is that these are 0 and we're setting them to the size/count of the start POST call to count them towards the batch. I'll add a comment to clarify.

sleroux commented:
Addressed the minor comments along with cleaning up the `addRecords` method. The problem with `accumulateRecords` before was that it wasn't very explicit. When we were able to accumulate, we returned nil. When we couldn't, we return an upload operation and forgot about the record which required the double-add logic in `addRecords`. I think this makes it a lot more clear by separating it out a bit. `accumulateRecord` now throws so now we `try` to accumulate a record. If we can't, the error contains the upload operation which we execute and chain an accumulateRecord call to for handling the missing record.

sleroux commented:
Also, regarding the OpaqueToken thing for the BatchToken, the server keeps sending it as a Number so whenever I try to parse it as a String from POSTResult it never works. If this is truly not a number, should it be coming from the server as a string instead of a number?

rnewman commented:
Might be a documentation error. Raise with @rfk?

rnewman commented on Sync/BatchingClient.swift:
```diff
@@ -0,0 +1,227 @@
+            return uploadOp >>> { self.accumulateOrUpload(record) }
+        } catch {
+            // Should never happen.
+            return succeed()
```
I'd return the error instead:

```
return deferMaybe(e)
```

rnewman commented on Sync/BatchingClient.swift:
```diff
@@ -0,0 +1,227 @@
+                    addToBatch(record)
+                    return
+                }
+                throw AccumulateRecordError.Full(uploadOp: self.postInBatch(token))
```
Can we invert these?

```
guard let token = self.batchToken else {
    guard addToPost(record) else {
        throw AccumulateRecordError.Full(uploadOp: self.start())
    }
    return
}

guard fitsInBatch(record) else {
    throw AccumulateRecordError.Full(uploadOp: self.commitBatch(token))
}

guard addToPost(record) else {
    throw AccumulateRecordError.Full(uploadOp: self.postInBatch(token))
}

addToBatch(record)
```

rnewman commented:
LGTM. Do one last pass to make sure all nits are covered (e.g., newline at EOF!), then let's test the shit out of this. 🍻

rfk commented:
> should it be coming from the server as a string instead of a number?

I agree, this is a server bug, it should be coming back as a string.

sleroux commented:
I've run through a bunch of manual testing scenarios on both production and staging and feeling pretty good about this. I'm going to go ahead and land this so we can test in our beta audience for the upcoming 6.0.
